### DISCUSSION TITLE
Could we improve error message for array/dictionary access?

### DISCUSSION THREAD

**Discussion by advent94** [godotengine/godot-proposals#9329] [Scripting]
I got interesting error message when I was trying to change element of read_only (implicitly casted from const) array.
It does tell me that I did "Invalid set on index[X]". This message is the same as if you try to access index that is out of range which can be confusing. 

Is there any reason not to explicitly state that user tried to access array index that is bigger than array size - 1 or that element he tried to change belongs to array that is declared as read only? 

It took me a while to find out what's wrong as I was doing some meta shenanigans, playing with array types. Since I encountered issue like this, I thought that maybe other people could be on the same boat and this can be improved, giving more clarity, information to fix small bugs like his.

**Comment by AThousandShips** [comment]
How did you confirm it was read-only? Can you show your code?

Edit: Can confirm with general read-only cases, both references to `const`, and explicitly read-only arrays, the message is a bit cryptic, and it's: `Invalid assignment of index '0' (on base: 'Array') with value of type 'int'.` and similar

This is specifically because of this code:
https://github.com/godotengine/godot/blob/fe01776f05b1787b28b4a270d53037a3c25f4ca2/core/variant/variant_setget.cpp#L652-L656

Which uses the same return, unsure how this could be easily fixed without re-writing a bunch of stuff here, but will take a look when I can

**Reply by advent94** [reply]
Constants
```GDScript 
const MIN_TIME_BETWEEN_UPDATES: float = 0.05
```

Update:
```GDScript
const MIN_TIME_BETWEEN_UPDATES: float = Constants.MIN_TIME_BETWEEN_UPDATES
```

Test
```GDScript 
const VALID_INTERVAL: float = Update.MIN_TIME_BETWEEN_UPDATES
const VALID_INTERVAL_ARRAY: Array[float] = [VALID_INTERVAL]

func should_make_deep_copy_of_existing_update():
	var update: Update = Update.new(VALID_INTERVAL_ARRAY)
		
	update.value[0] = 14.3
```

Update
```GDScript 
func _init(variant: Variant, copy: bool = false):
    ...
    # not applicable for valid array

    value = valid_interval_array(variant, copy)


func valid_interval_array(interval_array: Variant, copy: bool) -> Array[float]:
        ... 
        # not applicable for valid array

	for i in range(interval_array.size()):
                # this doesn't modify array itself in our reproduction case
		check_if_is_interval_valid(interval_array[i])

	...  
        # not applicable for valid array

	return interval_array
```

**Reply by advent94** [reply]
```GDScript
if update.value.is_read_only():
	print("READ_ONLY")
```

This would suffice to print but I found out that solution was to use 
```GDScript 
Array.duplicate() 
```
in constructor.

**Reply by AThousandShips** [reply]
Got a working improvement, will test it a bit more and then open a PR for it
